File: course_2/module_4/programming_assignment_4/solution.py
from concurrent.futures import ProcessPoolExecutor


class TwoSumSolver:
    def __init__(self, file_path, lower_bound, upper_bound):
        """
        Initializes the TwoSumSolver class with the file path and range bounds.

        Args:
            file_path (str): Path to the input file containing integers.
            lower_bound (int): The lower bound of the target range.
            upper_bound (int): The upper bound of the target range.
        """
        self.file_path = file_path
        self.lower_bound = lower_bound
        self.upper_bound = upper_bound
        self.numbers = set()

    def load_data(self):
        """
        Loads integers from the file into a set to ensure uniqueness.

        Raises:
            Exception: If the file cannot be read or parsed.
        """
        try:
            with open(self.file_path, 'r') as file:
                self.numbers = {int(line.strip()) for line in file}
        except Exception as e:
            raise Exception(f"Error loading data from file: {e}")

    def count_valid_targets(self, numbers_chunk):
        """
        Counts valid target values for a specific chunk of numbers.

        Args:
            numbers_chunk (list): A subset of the full numbers set.

        Returns:
            set: A set of valid target values for the given chunk.
        """
        targets = set()
        for num in numbers_chunk:
            for t in range(self.lower_bound, self.upper_bound + 1):
                complement = t - num
                if complement in self.numbers and complement != num:
                    targets.add(t)
        return targets

    def count_target_values(self):
        """
        Counts the number of distinct target values t in the range [lower_bound, upper_bound]
        such that t = x + y, where x and y are distinct numbers from the input.

        Returns:
            int: The count of distinct target values.
        """
        # Divide the numbers into chunks for parallel processing.
        num_processes = 8  # Number of processes to use.
        numbers_list = list(self.numbers)
        chunk_size = len(numbers_list) // num_processes
        chunks = [
            numbers_list[i * chunk_size:(i + 1) * chunk_size]
            for i in range(num_processes)
        ]
        # Ensure the last chunk includes all remaining numbers.
        if len(numbers_list) % num_processes != 0:
            chunks[-1].extend(numbers_list[num_processes * chunk_size:])

        # Use ProcessPoolExecutor for parallel computation.
        with ProcessPoolExecutor(max_workers=num_processes) as executor:
            results = executor.map(self.count_valid_targets, chunks)

        # Combine all results into a single set of targets.
        all_targets = set()
        for result in results:
            all_targets.update(result)

        return len(all_targets)


if __name__ == "__main__":
    # File path to the input data.
    file_path = "2sum.txt"
    lower_bound = -10000
    upper_bound = 10000

    # Initialize the TwoSumSolver.
    solver = TwoSumSolver(file_path, lower_bound, upper_bound)

    # Load data from the file.
    try:
        solver.load_data()
    except Exception as e:
        print(f"Error: {e}")
        exit(1)

    # Compute the number of distinct target values.
    result = solver.count_target_values()
    print(f"The number of target values in the range [{lower_bound}, {upper_bound}] is: {result}")

File: course_2/module_4/programming_assignment_4/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"sync"
)

type TwoSumSolver struct {
	filePath   string
	lowerBound int
	upperBound int
	numbers    map[int]struct{}
}

// NewTwoSumSolver creates a new instance of the TwoSumSolver.
func NewTwoSumSolver(filePath string, lowerBound, upperBound int) *TwoSumSolver {
	return &TwoSumSolver{
		filePath:   filePath,
		lowerBound: lowerBound,
		upperBound: upperBound,
		numbers:    make(map[int]struct{}),
	}
}

// LoadData reads integers from the file into a map for fast lookup.
func (solver *TwoSumSolver) LoadData() error {
	file, err := os.Open(solver.filePath)
	if err != nil {
		return fmt.Errorf("failed to open file: %w", err)
	}
	defer func() {
		if cerr := file.Close(); cerr != nil {
			fmt.Printf("Warning: failed to close file: %v\n", cerr)
		}
	}()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		num, err := strconv.Atoi(scanner.Text())
		if err != nil {
			return fmt.Errorf("failed to parse number: %w", err)
		}
		solver.numbers[num] = struct{}{}
	}

	if err := scanner.Err(); err != nil {
		return fmt.Errorf("error while reading file: %w", err)
	}

	return nil
}

// countTargetsInChunk calculates valid target values for a chunk of numbers.
func (solver *TwoSumSolver) countTargetsInChunk(numbersChunk []int, results chan<- map[int]struct{}, wg *sync.WaitGroup) {
	defer wg.Done()
	targets := make(map[int]struct{})

	for _, num := range numbersChunk {
		for t := solver.lowerBound; t <= solver.upperBound; t++ {
			complement := t - num
			if complement != num {
				if _, exists := solver.numbers[complement]; exists {
					targets[t] = struct{}{}
				}
			}
		}
	}
	results <- targets
}

// CountTargetValues calculates the number of distinct target values in parallel.
func (solver *TwoSumSolver) CountTargetValues() int {
	// Convert map keys to a slice for chunking.
	numbers := make([]int, 0, len(solver.numbers))
	for num := range solver.numbers {
		numbers = append(numbers, num)
	}

	// Divide the numbers into chunks for parallel processing.
	numWorkers := 8
	chunkSize := (len(numbers) + numWorkers - 1) / numWorkers
	results := make(chan map[int]struct{}, numWorkers)
	var wg sync.WaitGroup

	for i := 0; i < len(numbers); i += chunkSize {
		end := i + chunkSize
		if end > len(numbers) {
			end = len(numbers)
		}
		wg.Add(1)
		go solver.countTargetsInChunk(numbers[i:end], results, &wg)
	}

	// Close the results channel after all goroutines finish.
	go func() {
		wg.Wait()
		close(results)
	}()

	// Combine results from all goroutines.
	finalTargets := make(map[int]struct{})
	for partialTargets := range results {
		for t := range partialTargets {
			finalTargets[t] = struct{}{}
		}
	}

	return len(finalTargets)
}

func main() {
	// Path to the input file.
	filePath := "course_2/module_4/programming_assignment_4/2sum.txt" // Replace with the actual file path.
	lowerBound := -10000                                              // Lower bound of the range.
	upperBound := 10000                                               // Upper bound of the range.

	// Initialize the TwoSumSolver.
	solver := NewTwoSumSolver(filePath, lowerBound, upperBound)

	// Load data from the file.
	err := solver.LoadData()
	if err != nil {
		fmt.Printf("Error loading data: %v\n", err)
		return
	}

	// Compute the number of distinct target values.
	result := solver.CountTargetValues()
	fmt.Printf("The number of target values in the range [%d, %d] is: %d\n", lowerBound, upperBound, result)
}
